Objective-C interop doc updates (#6942)
dart-lang/site-www: src/content/interop/objective-c-interop.md

ffigen: "https://pub.dev/packages/ffigen"
example: "https://github.com/dart-lang/native/tree/main/pkgs/ffigen/example/objective_c"
ffigenapi: "https://pub.dev/documentation/ffigen/latest/ffigen"
ffigendoc: "https://github.com/dart-lang/native/blob/main/pkgs/ffigen/doc/README.md"
appledoc: "https://developer.apple.com/documentation"
---

Dart apps running on the [Dart Native platform][] (macOS or iOS)
can use `dart:ffi` and [`package:ffigen`][]
to call Objective-C and Swift APIs.

`dart:ffi` enables Dart code to interact with native C APIs.
Objective-C is based on and compatible with C,
so it's possible to interact with Objective-C APIs using only `dart:ffi`.
However, doing so requires significant boilerplate,
so you can use `package:ffigen` to automatically generate
the Dart FFI bindings for a given Objective-C API.
To learn more about FFI and interfacing with C code directly,
see the [C interop guide][].

You can generate Objective-C headers for Swift APIs,
enabling `dart:ffi` and `package:ffigen` to interact with Swift.

For more information about using FFIgen,
see the [FFIgen README][]
and the [additional documentation][].

> [Dart Native platform]: /overview#platform

Reviewer: If you have time, we should update the links in each section to look like this.

using [LLVM](https://llvm.org/),
so you'll need to install that first.
See [Installing LLVM]({{page.ffigen}}#installing-llvm)
from the FFIgen README for more details.

### Configuring FFIgen for Objective-C

First, add `package:ffigen` as a dev dependency:

```console
$ dart pub add --dev ffigen
```

Then, configure FFIgen to generate bindings for the
Objective-C header containing the API.
FFIgen can be configured via YAML or Dart code,
but Dart is preferred for new projects.
The YAML config will be deprecated in future versions of FFIgen.
Start by creating a `generate_code.dart` script somewhere in your package.
A good directory to create this file in is `my_package/tool`.

The `generate_code.dart` script should create an `FfiGenerator` object,
which will contain all our configuration options,
then call its `.generate()` method.

```dart
import 'package:ffigen/ffigen.dart';

final config = FfiGenerator(
);

void main() => config.generate();
```

First, we'll tell FFIgen where to find the API we're trying to
generate bindings for.
To do this, we set the `headers.entryPoints` option.

> For this example, we need to load `AVAudioPlayer.h`.

Reviewer 2: Generally the style guide suggests [addressing the reader as _you_](https://developers.google.com/style/person#address-the-reader-as-you):

```suggestion
First, you'll tell FFIgen where to find the API we're trying to
generate bindings for.
To do this, you set the `headers.entryPoints` option.

For this example, you'll load `AVAudioPlayer.h`.
```

When generating bindings for an Apple API,
exploring these directories is a great way to find
the right headers to pass to FFIgen.

```dart
import 'package:ffigen/ffigen.dart';

final config = FfiGenerator(
  headers: Headers(
    entryPoints: [
      Uri.file(
        '$macSdkPath/System/Library/Frameworks/AVFAudio.framework/Headers/AVAudioPlayer.h',
      ),
    ],
  ),
);

void main() => config.generate();
```

Next, we'll define the output file.
The main output of FFIgen is a single Dart file
containing bindings for the given inputs.
This file's location is defined by the `output.dartFile` option.

In some cases, FFIgen will also generate a `.m` file,
containing Objective-C code required for interop with the API.
This file will only be generated
if the API you're generating bindings for needs it
(e.g. if you're using blocks or protocols).
By default this file will have the same name as the Dart bindings,
but with `.m` at the end of the file name.
You can change its location with `output.objectiveCFile` option.
If FFIgen produces this file, you must compile it into your package,
otherwise you may get runtime exceptions to do with missing symbols.

> For this simple example we won't get a `.m` file.

Reviewer 2: ```suggestion
For this example, FFIgen doesn't generate a `.m` file.
```

Another important option is `output.preamble`.
This is simply text inserted at the top of the Dart output,
which can be useful for adding a license header
or disabling specific lints that FFIgen's output doesn't adhere to.
In this case we'll disable several lints.

> ```dart

Reviewer 2: If you want to draw attention to the output section focused on here, you can highlight the relevant lines:

````suggestion
```dart highlightLines=11-16
````

This file's location is defined by the `output.dartFile` option.

FFIgen sometimes generates a `.m` file,
containing Objective-C code required for interop with the API.
FFIgen generates this file only if the API requires it
(for example, if you're using blocks or protocols).
By default, this file has the same name as the Dart bindings,
but with `.m` at the end of the file name.
You can change its location with the `output.objectiveCFile` option.
If FFIgen produces this file, you must compile it into your package,
otherwise you might get runtime exceptions relating to missing symbols.
For this example, FFIgen doesn't generate a `.m` file.

Use the `output.preamble` option to insert text at the top of the Dart output,
such as license headers or lint directives.
In this case, disable several lints:

```dart highlightLines=11-16
import 'package:ffigen/ffigen.dart';

final config = FfiGenerator(
  headers: Headers(
    entryPoints: [
      Uri.file(
        '$macSdkPath/System/Library/Frameworks/AVFAudio.framework/Headers/AVAudioPlayer.h',
      ),
    ],
  ),
  output: Output(
    dartFile: Uri.file('avf_audio_bindings.dart'),
    preamble: '''

> // ignore_for_file: camel_case_types, non_constant_identifier_names, unused_element, unused_field, void_checks, annotate_overrides, no_leading_underscores_for_local_identifiers, library_private_types_in_public_api

Author: Not sure. This code is the same as what I have in `ffigen/example/objective_c`, and I needed it there to avoid analysis issues on CI. But maybe you don't need it when you're doing this stuff locally. Does `dart analyze` run without errors/warnings in your [repo](https://github.com/goderbauer/scratchpad/tree/main/objcxp)?

   There is an [open feature request][] to enable isolates to be
   pinned to specific threads.
2. While FFIgen supports converting
   Dart functions to Objective-C blocks,
   most Apple APIs don't make any guarantees about
   which thread a callback will run on.
3. Most APIs that involve UI interaction
   can only be called on the main thread,
   also called the platform thread in Flutter.
4. Many Apple APIs are [not thread safe][].

The first two points mean that a block created in one isolate
might be invoked on a thread running a different isolate,
or no isolate at all.
Depending on the type of block you are using,
this could cause your app to crash.
When a block is created, the isolate it was created in is its owner.
Blocks created using `FooBlock.fromFunction`
must be invoked on the owner isolate's thread,
otherwise they will crash.
Blocks created using `FooBlock.listener` or `FooBlock.blocking`
can be safely invoked from any thread,
and the function they wrap will (eventually) be invoked
inside the owner isolate,
though these constructors are only supported for blocks that return `void`.
`FooBlock.blocking` may add support for non-`void` return values in future,

> if there is user demand for it.

Reviewer 2: ```suggestion
If there is demand for it,
`FooBlock.blocking` might add support for
non-`void` return values in the future.
```

The third point means that directly calling some Apple APIs
using the generated Dart bindings might be thread unsafe.
This could crash your app, or cause other unpredictable behavior.
In recent versions of Flutter, the main isolate runs on the platform thread,
so this isn't an issue when invoking these thread-locked APIs
from the main isolate.
If you need to invoke these APIs from other isolates,

> or you need to support older versions of flutter,

Reviewer 2: ```suggestion
or you need to support older versions of Flutter,
```
